Sources/Crust/Parser.swift

```diff
@@ -42,4 +50,543 @@ public class Parser {
+  func parseQualifiedName() throws -> QualifiedNameSyntax {
+    var pieces = [QualifiedNamePieceSyntax]()
+    while let piece = try? parseQualifiedNamePiece() {
+      pieces.append(piece)
+    }
+
+    guard !pieces.isEmpty else {
+      throw ParseError.failedParsingName
+    }
+
+    return QualifiedNameSyntax(elements: pieces)
+  }
+
+  func parseQualifiedNamePiece() throws -> QualifiedNamePieceSyntax {
+    guard case .identifier(_) = self.peek() else {
+      throw ParseError.unexpectedToken(self.currentToken!)
+    }
+
+    let name = self.currentToken!
+    self.advance()
+    let period = (peek() == .period) ? try? consume(.period) : nil
+    return QualifiedNamePieceSyntax(name: name, trailingPeriod: period)
+  }
+
+  func parseIdentifierList() throws -> IdentifierListSyntax {
+    var pieces = [TokenSyntax]()
+    while let piece = try? parseQualifiedNamePiece() {
+      pieces.append(piece.name)
+    }
+    return IdentifierListSyntax(elements: pieces)
+  }
+}
+
+extension Parser {
+  func parseDeclList() -> DeclListSyntax {
+    var pieces = [DeclSyntax]()
+    while let piece = try? parseDecl() {
+      pieces.append(piece)
+    }
+    return DeclListSyntax(elements: pieces)
+  }
+
+  func parseDecl() throws -> DeclSyntax {
+    return try split({
+      return try self.parseDataDeclaration()
+    }, {
+      return try self.parseModule()
+    }, {
+      return try self.split({
+        return try self.parseOpenImportDecl()
+      }, {
+        return try self.parseImportDecl()
+      })
+    }, {
+      return try self.parseFunctionDecl()
+    })
+  }
+
+  func parseModule() throws -> ModuleDeclSyntax {
+    let moduleKw = try consume(.moduleKeyword)
+    let moduleId = try parseQualifiedName()
+    let paramList = parseTypedParameterList()
+    let whereKw = try consume(.whereKeyword)
+    let leftBrace = try consume(.leftBrace)
+    let declList = parseDeclList()
+    let rightBrace = try consume(.rightBrace)
+    return ModuleDeclSyntax(
+      moduleToken: moduleKw,
+      moduleIdentifier: moduleId,
+      typedParameterList: paramList,
+      whereToken: whereKw,
+      leftBraceToken: leftBrace,
+      declList: declList,
+      rightBraceToken: rightBrace
+    )
+  }
+
+  func parseOpenImportDecl() throws -> OpenImportDeclSyntax {
+    let openTok = try consume(.openKeyword)
+    let importTok = try consume(.importKeyword)
+    let ident = try parseQualifiedName()
+    return OpenImportDeclSyntax(
+      openToken: openTok,
+      importToken: importTok,
+      importIdentifier: ident
+    )
+  }
+
+  func parseImportDecl() throws -> ImportDeclSyntax {
+    let importTok = try consume(.importKeyword)
+    let ident = try parseQualifiedName()
+    return ImportDeclSyntax(importToken: importTok, importIdentifier: ident)
+  }
+}
+
+extension Parser {
+  func parseRecordDecl() throws -> RecordDeclSyntax {
+    let recordTok = try consume(.recordKeyword)
+    let recName = try parseQualifiedNamePiece().name
```

Comment:
Won’t this break the fidelity contract? We throw away the parent trivia. We need to update the grammar if this is wrong.

Reply:
Just lazy.  I'll fix it